Portable BPF map implementation packages

tommyp1ckles opened the pull request:
Refactoring various map types in order to make type declaration code portable.
Currently, when trying to import/build against map packages (ex. `pkg/maps/lxcmap`) it is not possible to build for non linux platforms.
This is because these packages import pkg/bpf which relies on Linux specific bpf C libraries.

To resolve this, this PR separates struct declarations used for implementing the bpf.MapKey/bpf.MapValue types from their implementation packages. Data struct declarations are moved to `./types` subpackage for each map type (i.e. `pkg/maps/lxcmap/types`).

Implementation is then added by aliasing the type in the original map package. For example:

**note**: code gen headers, such as deepcopy-gen should remain in the implementation package, not with the types type declarations.

`pkg/maps/ipcache/types/ipcache.go`
```go
type Key struct {
	Prefixlen uint32 `align:"lpm_key"`
	Pad1      uint16 `align:"pad1"`
	Pad2      uint8  `align:"pad2"`
	Family    uint8  `align:"family"`
	// represents both IPv6 and IPv4 (in the lowest four bytes)
	IP types.IPv6 `align:"$union0"`
}
```

and `pkg/maps/ipcache/ipcache.go`

```go
// Key implements the bpfTypes.MapKey interface.
// Must be in sync with struct ipcache_key in <bpf/lib/maps.h>
// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf/types.MapKey
type Key ipcacheTypes.Key

func (k Key) NewValue() bpfTypes.MapValue { return &RemoteEndpointInfo{} }

// etc ...
```

To ensure that this constraint remains true, added a CI check that will build all go packages in Cilium named "types" against non Linux GOOS targets to ensure that they are portable (also ensures portability against unix specific libraries such `sys/unix`).

```release-note
Refactor pkg/maps declaration types to be portable
```

tommyp1ckles committed:
ci: add lint check to ensure that */types packages are portable.

Any Go package ending with types will be checked that it compiles on
common, non-linux, platforms.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
types: implement json marshal/unmarshal for proto/ip/mac addr types.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

nbusseneau commented on .github/workflows/lint-go.yaml:
```diff
@@ -86,6 +86,23 @@ jobs:
+      - name: Install Go
+        uses: actions/setup-go@c4a742cab115ed795e34d4513e2cf7d472deb55f
+        with:
+          go-version: 1.19.2
```
[Nit] Seems like `go-version` has been updated to 1.19.2 in other jobs.

nbusseneau reviewed (approved):
See nit above. Otherwise workflow changes LGTM.

jrajahalme commented on pkg/maps/lbmap/types/lbmap.go:
```diff
@@ -0,0 +1,3 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package types
```
Seems like this file is not needed?

jrajahalme commented on pkg/maps/lbmap/types/maglev.go:
```diff
@@ -0,0 +1,3 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package types
```
Seems like this file is not needed?

jrajahalme commented on pkg/maps/lbmap/types/maglev_inner_map.go:
```diff
@@ -0,0 +1,3 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package types
```
Seems like this file is not needed?

jrajahalme commented on pkg/maps/lbmap/types/maglev_outer_map.go:
```diff
@@ -0,0 +1,3 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package types
```
Seems like this file is not needed?

jrajahalme commented on pkg/maps/lbmap/types/types.go:
```diff
@@ -0,0 +1,3 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package types
```
Seems like this file is not needed?

tommyp1ckles replied:
Woops, yup, this was partially refactored using a program, need to remove no-op cases where it still wrote the file

jrajahalme reviewed (approved):
LGTM apart from a few unnecessary files being added?

tommyp1ckles committed:
pkg/bpf/types: fix bpf pkg tests.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/bpf: refactor pkg/bpf MapKey/MapValue and map types.

Seperates all map data struct type declarations into packages that do
not need to link against Linux specific C libraries, such as pkg/bpf.
Moving bpf.MapKey and bpf.MapValue to a separate sub package in order
to split implementation out of the interfaces.
Also moving type declarations of all implementation structs for
MapKey/MapValue into sub packages "types" and replacing implementation
types with type aliases.

Together these changes allow for portable use of the pkg/map key/value
types externally.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

pippolo84 commented on pkg/maps/README:
```diff
@@ -0,0 +1,32 @@
+# Maps Implementations
+
+This directory contains implementation details for specific BPF Maps used by Cilium.
+Code in the primary directory (e.x. ./lxamp) should contain all code dealing with implementing
```
typo: `./lxamp` -> `./lxcmap`

pippolo84 reviewed (approved):
LGTM minus a small typo. Thanks! :rocket:

julianwiedmann reviewed (commented):
Looks like datapath was already covered by Jussi - so dropping myself to not hold this up further.

Just two small comments:

- `ci: add lint check to ensure that */types packages are portable.` also touches `pkg/maps/ctmap/`. That part probably belongs into the initial patch?
- should `pkg/bpf/types: fix bpf pkg tests.` also be squashed into the initial patch?
